File: poh2/poh2.java
//import java.util.*;
class Main{
	static final int SIZE=9999999;
	static byte[] z=new byte[SIZE];
	static int input_count=0;

	static int get(){
		int r=0;
		for(;'0'<=z[input_count]&&z[input_count]<='9';)r=r*10+z[input_count++]-'0';
		input_count++;
		return r;
	}
	static int getC(){
		return z[input_count++];
	}
	public static void main(String[]args){try{
		System.in.read(z,0,SIZE);
		//Scanner cin=new Scanner(System.in);
		int h,w,s,t,i,j,k;
		//h=cin.nextInt();
		//w=cin.nextInt();
		//cin.nextLine();
		h=get();
		w=get();
		int[][][] a=new int[h][w][h];
		int[][] accum=new int[h][w+1];
		for(i=0;i<h;i++){
			//String line=cin.nextLine();
			for(j=0;j<w;j++)a[i][j][0]=(getC()-'0')^1;
			getC();
		}
		for(i=0;i<h;i++){
			accum[0][a[i][0][0]]++;
			for(j=1;j<w;j++){
				if(a[i][j][0]>0){
					a[i][j][0]+=a[i][j-1][0];
					accum[0][a[i][j][0]]++;
				}
			}
		}
		for(i=1;i<h;i++)for(j=0;j<w;j++)for(k=1;k<=i;k++){
			a[i][j][k]=Math.min(a[i-1][j][k-1],a[i][j][k-1]);
			if(a[i][j][k]==0)break;
			accum[k][a[i][j][k]]++;
		}
		for(i=0;i<h;i++)for(j=w-1;j>=0;j--)accum[i][j]+=accum[i][j+1];
		//k=cin.nextInt();
		k=get();
		for(;k>0;k--){
			//s=cin.nextInt();
			//t=cin.nextInt();
			s=get();
			t=get();
			System.out.println(s<=h&&t<=w?accum[s-1][t]:0);
		}
	}catch(java.io.IOException e){}}
}

File: poh2/poh2.kt
import java.util.*
val SIZE=9999999
val z=ByteArray(SIZE)
var input_count=0

fun get():Int{
	var r:Int=0
	while('0'.toByte()<=z[input_count]&&z[input_count]<='9'.toByte())r=r*10+z[input_count++]-'0'.toByte()
	input_count++
	return r
}
fun getC():Int{
	return z[input_count++].toInt()
}
fun main(args: Array<String>){try{
	System.`in`.read(z,0,SIZE)
	var h:Int=get()
	var w:Int=get()
	var a=Array(h,{Array(w,{IntArray(h)})})
	var accum=Array(h,{IntArray(w+1)})
	for(i in 0..h-1){
		for(j in 0..w-1){
			a[i][j][0]=(getC()-'0'.toInt()) xor 1
		}
		getC();
	}
	for(i in 0..h-1){
		accum[0][a[i][0][0]]++
		for(j in 1..w-1){
			if(a[i][j][0]>0){
				a[i][j][0]+=a[i][j-1][0]
				accum[0][a[i][j][0]]++
			}
		}
	}
	for(i in 1..h-1)for(j in 0..w-1)for(k in 1..i){
		a[i][j][k]=Math.min(a[i-1][j][k-1],a[i][j][k-1])
		if(a[i][j][k]==0)break
		accum[k][a[i][j][k]]++
	}
	for(i in 0..h-1){
		var j:Int=w-1
		while(j>=0){
			accum[i][j]+=accum[i][j+1]
			j-=1
		}
	}
	var k:Int=get()
	while(k>0){
		var s:Int=get()
		var t:Int=get()
		println(if(s<=h&&t<=w)accum[s-1][t] else 0)
		k-=1
	}
}catch(e:java.io.IOException){}}
